consumer/kafka/src/main/java/com/spotify/heroic/consumer/kafka/ConsumerThread.java

```diff
@@ -83,23 +82,38 @@ public ConsumerThread(
     public void run() {
         log.info("{}: Starting thread", name);
 
-        active.incrementAndGet();
+        threadIsStarting();
 
         try {
             guardedRun();
         } catch (final Throwable e) {
             log.error("{}: Error in thread", name, e);
-            active.decrementAndGet();
+            threadIsStopping();
             stopFuture.fail(e);
             return;
         }
 
         log.info("{}: Stopping thread", name);
-        active.decrementAndGet();
+        threadIsStopping();
         stopFuture.resolve(null);
         return;
     }
 
+    private void threadIsStarting() {
+        // Increment counter and report atomically
+        synchronized (active) {
```

Comment:
you can avoid synchronizing by making the interaction with the reporting framework:
```java
reporter.incrementActiveThreads()
/*  */
reporter.decrementActiveThreads()
```

And have the `active-threads` just being a counter.

Because the reporter is shared across threads, all synchronization has to happen there. There are no guarantees an arbitrary caller will yield to this lock.

I also realize it seems compelling to have the `active` counter being in-sync between what's being reported and what is not. They are used for two different things, so it's not a big deal if they are not.